api/v1beta1/conversion_test.go

```diff
@@ -18,75 +20,83 @@ func TestCompatibility(t *testing.T) {
 		t.Parallel()
 
 		scheme := newScheme(t)
-		f := newCompatibilityTestFuzzer(t, scheme)
-
-		for i := 0; i < 10000; i++ {
-			var v1beta1Cluster1, v1beta1Cluster2 MySQLCluster
-			var v1beta2Cluster mocov1beta2.MySQLCluster
-			f.Fuzz(&v1beta1Cluster1)
-
-			var tmp1, tmp2 mocov1beta2.MySQLCluster
-
-			if err := scheme.Convert(v1beta1Cluster1.DeepCopy(), &tmp1, nil); err != nil {
-				t.Fatal(err)
-			}
-			if err := scheme.Convert(&tmp1, &v1beta2Cluster, nil); err != nil {
-				t.Fatal(err)
-			}
-			if err := scheme.Convert(&v1beta2Cluster, &tmp2, nil); err != nil {
-				t.Fatal(err)
-			}
-			if err := scheme.Convert(&tmp2, &v1beta1Cluster2, nil); err != nil {
-				t.Fatal(err)
-			}
-
-			if diff := cmp.Diff(v1beta1Cluster1, v1beta1Cluster2, cmpopts.EquateEmpty()); diff != "" {
-				t.Fatalf("compatibility error case #%d (-want +got):\n%s", i, diff)
-			}
+
+		var v1beta1Cluster1, v1beta1Cluster2 MySQLCluster
+		var v1beta2Cluster mocov1beta2.MySQLCluster
+
+		obj, err := roundtrip.CompatibilityTestObject(scheme, GroupVersion.WithKind("MySQLCluster"), fillFuncs())
+		if err != nil {
+			t.Fatal(err)
+		}
+		v1beta1Cluster1 = *obj.(*MySQLCluster)
+
+		var tmp1, tmp2 mocov1beta2.MySQLCluster
+
+		if err := scheme.Convert(v1beta1Cluster1.DeepCopy(), &tmp1, nil); err != nil {
+			t.Fatal(err)
+		}
+		if err := scheme.Convert(&tmp1, &v1beta2Cluster, nil); err != nil {
+			t.Fatal(err)
+		}
+		if err := scheme.Convert(&v1beta2Cluster, &tmp2, nil); err != nil {
+			t.Fatal(err)
+		}
+		if err := scheme.Convert(&tmp2, &v1beta1Cluster2, nil); err != nil {
+			t.Fatal(err)
+		}
```

Comment:
Could you tell me why we converted it four times? Sorry, I forgot :(

Reply:
Certainly no need for 4 conversions...

When I implemented conversion in the past, I tried to prepare internal objects in addition to v1beta1 and v1beta2.
In the controller logic, using an internal objcet has the advantage of not having to change the api version used by the controller when the api version increases in the future.

I prepare an internal object, you will need to do the conversion 4 times:

* v1beta1 -> internal -> v1beta2 -> internal -> v1beta1

However, the implementation became so complex that we decided not to adopt internal objects.
I had forgotten to make the fix to the test at that time.

Reply:
I will modify it to the proper conversion logic.
Thanks for review 🙏

Reply:
fixed: https://github.com/cybozu-go/moco/pull/429/commits/27856251cf9b6b066f98a58853b213e0d9e68767